test/unit/sqs/index.test.ts

```diff
@@ -22,140 +23,296 @@ describe("SqsQueue tests", () => {
+
+			queue.start();
+
+			queue.sendMessage(testPayload);
 		});
 
-		queue.stop();
+		test("Message received with delay", (done: DoneCallback) => {
 
-		//delaying to make sure all asynchronous invocations inside the queue will be finished and it will stop
-		await queue.waitUntilListenerStopped();
+			const testPayload = generatePayload();
+			const receivedTime = {time: new Date().getTime()};
 
-		queue.start();
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
 
-		queue.sendMessage(testPayload);
-	});
+				if(context.payload.msg !== testPayload.msg) {
+					//discard message if it is not our test message
+					return;
+				}
+
+				try {
+					context.log.info("hi");
+					const currentTime = new Date().getTime();
+					expect(currentTime - receivedTime.time).toBeGreaterThanOrEqual(1000);
+					done();
+				} catch (err) {
+					done(err);
+				}
+			});
+			queue.sendMessage(testPayload, 1);
+		});
+
+		test("Message gets executed exactly once", (done: DoneCallback) => {
+
+			const testPayload = generatePayload();
+			const testData: { messageId: undefined | string } = {messageId: undefined};
+
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
+
+				if(context.payload.msg !== testPayload.msg) {
+					//discard message if it is not our test message
+					return;
+				}
+
+				try {
+					expect(context.payload).toStrictEqual(testPayload);
+
+					if (!testData.messageId) {
+						testData.messageId = context.message.MessageId;
+					} else if (testData.messageId === context.message.MessageId) {
+						done.fail("Message was received more than once");
+					} else {
+						done.fail("Different message on the tests queue");
+					}
+
+				} catch (err) {
+					done(err);
+				}
+			});
+			queue.sendMessage(testPayload);
+
+			//code before the pause
+			setTimeout(function () {
+				if (testData.messageId) {
+					done();
+				} else {
+					done.fail("No message was received");
+				}
+			}, 3000);
+
+		});
+
+		test("Messages are not processed in parallel", async (done: DoneCallback) => {
+
+			const testPayload = generatePayload();
+			const receivedTime = {time: new Date().getTime(), counter: 0};
+
+			mockRequestHandler.handle.mockImplementation(async (context: Context<TestMessage>) => {
+				try {
 
-	test("Message received with delay", (done: DoneCallback) => {
+					if(context.payload.msg !== testPayload.msg) {
+						//discard message if it is not our test message
+						return;
+					}
 
-		const testPayload = { msg: "Hi4" };
-		const receivedTime = { time: new Date().getTime() };
+					if (receivedTime.counter == 0) {
+						receivedTime.counter++;
+						context.log.info("Delaying the message");
+						await delay(1000);
+						context.log.info("Message processed after delay");
+						return;
+					}
+
+					const currentTime = new Date().getTime();
+					expect(currentTime - receivedTime.time).toBeGreaterThanOrEqual(1000);
```

Comment:
This will be flaky.

Reply:
This needs to be paid attention.  This situation has already happened at least once before with one of your previous PRs and we had to fix it because it kept failing sporadically for others.  You cannot use timers in tests, it's extremely unpredictable.  This is the reason [jest has a whole section dedicated to handling these time based issues](https://jestjs.io/docs/timer-mocks).

Reply:
Following up on the discussion we had F2F. That's the problem you are talking about, which we address in the past
https://github.com/atlassian/github-for-jira/pull/853/files

Flakiness had nothing to do with timeouts